src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -0,0 +1,502 @@
+      var es = entries(current, poly)
+      if (es != null && (es(idx) eq entry)) current
+      else {
+        val result = 
+          if (es == null) {
+            es = Array.fill(poly.paramNames.length)(initial)
+            updateEntries(current, poly, es)
+          }
+          else if (es ne entries(prev, poly))
+            current // can re-use existing entries array.
+          else {
+            es = es.clone
+            updateEntries(current, poly, es)
+          }
+        es(idx) = entry
+        result
+      }
+    }
+    
+    def update(prev: OrderingConstraint, current: OrderingConstraint, 
+        param: PolyParam, entry: T)(implicit ctx: Context): OrderingConstraint =
+      update(prev, current, param.binder, param.paramNum, entry)
+      
+    def map(prev: OrderingConstraint, current: OrderingConstraint, 
+        poly: PolyType, idx: Int, f: T => T)(implicit ctx: Context): OrderingConstraint = 
+     update(prev, current, poly, idx, f(apply(current, poly, idx)))
+
+    def map(prev: OrderingConstraint, current: OrderingConstraint, 
+        param: PolyParam, f: T => T)(implicit ctx: Context): OrderingConstraint = 
+      map(prev, current, param.binder, param.paramNum, f)
+  }
+
+  val boundsLens = new ConstraintLens[Type] {
+    def entries(c: OrderingConstraint, poly: PolyType): Array[Type] = 
+      c.boundsMap(poly)
+    def updateEntries(c: OrderingConstraint, poly: PolyType, entries: Array[Type])(implicit ctx: Context): OrderingConstraint = 
+      newConstraint(c.boundsMap.updated(poly, entries), c.lowerMap, c.upperMap)
+    def initial = NoType
+  }
+  
+  val lowerLens = new ConstraintLens[List[PolyParam]] {
+    def entries(c: OrderingConstraint, poly: PolyType): Array[List[PolyParam]] = 
+      c.lowerMap(poly)
+    def updateEntries(c: OrderingConstraint, poly: PolyType, entries: Array[List[PolyParam]])(implicit ctx: Context): OrderingConstraint = 
+      newConstraint(c.boundsMap, c.lowerMap.updated(poly, entries), c.upperMap)
+    def initial = Nil
+  }
+
+  val upperLens = new ConstraintLens[List[PolyParam]] {
+    def entries(c: OrderingConstraint, poly: PolyType): Array[List[PolyParam]] = 
+      c.upperMap(poly)
+    def updateEntries(c: OrderingConstraint, poly: PolyType, entries: Array[List[PolyParam]])(implicit ctx: Context): OrderingConstraint = 
+      newConstraint(c.boundsMap, c.lowerMap, c.upperMap.updated(poly, entries))
+    def initial = Nil
+  }
+}
+
+import OrderingConstraint._
+
+/** Constraint over undetermined type parameters that keeps separate maps to 
+ *  reflect parameter orderings.
+ *  @param boundsMap a map from PolyType to arrays.
+ *               Each array contains twice the number of entries as there a type parameters
+ *               in the PolyType. The first half of the array contains the type bounds that constrain the
+ *               polytype's type parameters. The second half might contain type variables that
```

Comment:
Why use half of an array for something and the other half for something else, instead of just using a map of PolyType to pair of arrays or two maps of PolyType to arrays? This is worth documenting.